source/_posts/2015-10-17-headless-firefox-in-ubuntu-on-virtual-box-for-cucumber-testing.markdown

```diff
@@ -0,0 +1,80 @@
+---
+layout: post
+title: "Headless Firefox in Ubuntu on VirtualBox for Cucumber testing"
+date: 2015-10-17 14:12:56 +0200
+comments: true
+author: Ivica Lakatoš
+categories: 
+  - Ruby and Rails
+  - Testing
+tags: 
+  - ruby
+  - rails
+  - vagrant
+  - virtualbox
+  - ubuntu
+  - cucumber
+  - selenium-webdriver
+published: true
+---
+
+If you use [Vagrant](http://www.vagrantup.com/downloads.html), [VirtualBox](https://www.virtualbox.org/) and Ubuntu to build your Rails apps and you want to test it with Cucumber scenarios, this is right post for you. By default Vagrant and VirtualBox use Ubuntu without an X server and GUI.
+
+Everything goes well until you need `@javascript` flag for your cucumber scenario. `@javascript` uses a javascript-aware system to process web requests (e.g. Selenium) instead of the default (non-javascript-aware) webrat browser.
+
+### Install Mozilla Firefox
+
+Selenium WebDriver is flexible and let you run selenium headless in servers with no display. But in order to run, Selenium needs to launch a browser. If there are no display to the machine, the browsers are not launched. So in order to use selenium, you need to fake a display and let selenium and the browser thinks they are running in a machine with a display.
+
+Install latest version of Mozilla Firefox:
+
+`sudo apt-get install firefox`
+
+Since Ubuntu is running without a X server Selenium cannot start Firefox because it requires an X server.
+
+### Setting up virtual X server
+
+Virtual X server is required to make browsers run normally by making them believe there is a display available, although it doesn't create any visible windows. 
+<!--more-->
+Xvfb (X Virtual FrameBuffer) works fine for this. Xvfb lets you run X-Server in machines with no display devices. 
+
+Install xvfb on ubuntu:
+
+`sudo apt-get install xvfb`
+
+Lets run the Xvfb service in a display number which is less likely to clash even if you add a display at later stage. Display 10 will do fine.
```

Comment:
@Ikera 
if you add a display at later stage -> if you add a display at **a** later stage